src/psyclone/psyir/transformations/omp_target_trans.py

```diff
@@ -88,6 +93,53 @@ class OMPTargetTrans(RegionTrans):
     '''
     excluded_node_types = (CodeBlock, )
 
+    def _add_asynchronicity(self, nodes: List[Node], instance: Directive):
+        '''
+        TODO
+        '''
+        next_depend = self._find_next_dependency(nodes, instance)
+
+        # If find_next_dependency returns False, then this loop is its own
+        # next dependency so we can't add an asynchronous clause.
+        if not next_depend:
+            return
+        # If find next_dependency returns True there is no follow up
+        # dependency, so we just need a barrier at the end of the containing
+        # Routine.
+        if next_depend is True:
+            # Add nowait to the instance.
+            instance.nowait = True
+            # Add a barrier to the end of the containing Routine if there
+            # isn't one already.
+            containing_routine = instance.ancestor(Routine)
+            # Check barrier that corresponds to self.omp_directive and add the
+            # correct barrier type
+            if not isinstance(containing_routine.children[-1],
+                              OMPTaskwaitDirective):
+                containing_routine.addchild(OMPTaskwaitDirective())
+            return
+
+        # Otherwise we have the next dependency and we need to find where the
+        # correct place to place the preceding barrier is. Need to find a
+        # guaranteed control flow path to place it.
+
+        # Find the deepest schedule in the tree containing both.
+        sched = next_depend.ancestor(Schedule)
```

Comment:
This bit looks like it *is* a duplicate of the one in omp_loop_trans?

Reply:
It is - again I'm not sure a good ancestor to whom it would belong (since the mixin is not specific to OpenMP is the idea).

Reply:
Yeah. I think we'll leave this issue of duplication until we're a bit further on with this functionality - it may become clearer then.